man/sms.8
.TH SMS 8 "Mon Jul 14 2003" "" "Network UPS Tools (NUT)"
.SH NAME
sms \- Driver for Brazilian SMS UPS equipment
.SH NOTE
This driver is obsolete and has been replaced by the \fBmegatec\fR(8)
driver. It will be removed somewhere in the near future.

This man page only documents the hardware\(hyspecific features of the
sms driver.  For information about the core driver, see
\fBnutupsdrv\fR(8).

.SH SUPPORTED HARDWARE
This driver has been tested with :
  

Manager III   650 VA

Manager III  1300 VA\fR  

Manager III   650 VA sinusoidal

Manager III  1300 VA sinusoidal

Sinus Single 2000 VA sinusoidal on\(hyline 

Please report another SMS working models, or send\(hyme the driver 
detection string.  

.SH EXTRA ARGUMENTS

This driver does not support any extra settings in the
\fBups.conf\fR(5).

.SH UPSCMD


test.battery.stop \(hy Cancel battery test

test.battery.start \(hy Run battery test until battery is low

test.failure.start \(hy Simulate power failure

shutdown.return \(hy Shut down in .3 minutes and restart in .3 minutes

shutdown.stayoff \(hy Shut down in .3 minutes and do not return

shutdown.stop \(hy Stop a shutdown command

reset.watchdog \(hy Shutdown in 5 minutes and restart in 3 minutes.  Call
this with upscmd in crontab every minute.


.SH AUTHOR
Marcio Gomes <tecnica@microlink.com.br>

.SH SEE ALSO

.SS The core driver:
\fBnutupsdrv\fR(8)

.SS Internet resources:
The NUT (Network UPS Tools) home page: http://www.networkupstools.org/


NEWS
If you're upgrading from an earlier version, see the UPGRADING file.

---------------------------------------------------------------------------
Release notes for NUT 2.0.5 - what's new since 2.0.4:

 This release is a backport of the development version.  Many changes
 have already been backported previously.  Thus it is more a
 synchronisation release, though it includes many bugfixes and support
 for new models.

 - support for new devices: APC Smart-UPS with 6TI firmware; Belkin
   Small Enterprise F6C1500-TW-RK; Compaq R3000 XR, R5500 XR; Cyber
   Power 550SL, 725SL, 685AVR, 800AVR, 1200AVR, AE550; Eltek; Inform
   GUARD; Microsol Rhino; Opti-UPS PowerES 420E; PowerMan RealSmart,
   BackPro; Powerware PW9315 3-phase; SOLA 305; Tripp-Lite
   SMART550USB, SMART2200RMXL2U, OMNI1000LCD, OMNI900LCD, OMNI650LCD,
   1500 LCD, AVR550U; Viewsonic PowerES 420E.
 - bcmxcp: added 3-phase support
 - megatec: better hardware support, more instant commands
 - mge-hid: support more instant commands
 - newhidups: fixed APC and Tripp Lite bugs, various memory bugs,
   improved report buffering, improved Solaris support, added
   '-x explore' option for easy diagnosis of new devices
 - solis: shutdown programming, support new cables, Solaris support
 - tripplite_usb: updated SMARTPRO support, fixed OL/OB reporting, 
   better error handling, some memory bugs
 - new dummy-ups driver simulator
 - added HTML interface for access to CGI scripts

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.0.4 - what's new since 2.0.3:

 - The newhidups critical bug (segmentation fault) has been fixed. It has
   also received some more care, like buxfixes and new models support and
   enhancement for Solaris.
   [Peter Selinger and Arnaud Quette]

 - A bug has been fixed in NUT core to support resuming from suspend-to-disk.
   This should also fix other similar issues, like time synchronisation
   through the NTP - Network Time Protocol.
   [Arjen de Korte]

 - The mge-shut driver now better detects the Low Battery status, support
   new models and fixes some wrong status and data. It also fixes some
   issue where the UPS wasn't restarting (refer to mge-shut manpage).
   [Arnaud Quette]

 - The genericups custom configuration through ups.conf is working again
   [Arjen de Korte]

 - The genericups driver type 22 also support CyberPower 725SL
   (and maybe others SL models)
   [David Kaufman]

 - The new megatec driver, which will replace a bunch of drivers by nut 2.2
   (refer to docs/megatec.txt and UPGRADING) has been backported from the
   trunk (Development tree). The powermust driver has also received some
   attention.
   [Carlos Rodrigues]

 - The new rhino driver was added to support Microsol Rhino UPS hardware
   The solis has also been improved for solaris compatibility, and
   internal / external shutdown programming. solis can now save external
   shutdown programming to ups, and support new cables for solis 3
   [Silvino B. MagalhÃ£es]

 - Several fixes and improvements have been made to upsrw, upsset,
   cpsups, tripplite_usb and the FAQ.
   [Arjen de Korte and Charles Lepple]

---------------------------------------------------------------------------
Release notes for NUT 2.0.3 - what's new since 2.0.2:

 - The recent and major newhidups changes have been backported from the
   Development tree. It now:
   - supports models from MGE UPS SYSTEMS, APC and Belkin. Mustek and Unitek
     units are also recognized for development purpose,
   - handles better device reopening, after a disconnection,
   - handles multiple devices, with several parameters to find the right UPS.
   [Peter Selinger, Charles Lepple and Arnaud Quette]

 - The bcmxcp_usb driver has been added to support Powerware USB units.
   [Wolfgang Ocker and Kjell Claesson]

 - The tripplite_usb driver has been added to support Tripp Lite USB units.
   [Charles Lepple]

 - The sec driver is back as gamatronic
   [Gamatronic, Nadav Moskovitch]

 - The genericups driver has received official care from Gamatronic
   to add support for the Gamatronic UPS with alarm interface.
   [Gamatronic, Nadav Moskovitch]

 - The powermust driver now supports Soyntec Sekury C 500 and C 800 units.
   [Hanno Borns]

 - The mge-shut driver has received a bit of attention too, and enhance
   ups.model retrieval for some specific case (release 0.65)

 - The drivers don't change to the "statepath" directory anymore at
   initialisation time if called using -k. This avoid unneeded
   failure to poweroff the UPS if /var is already unmounted.
   [Gaspar Bakos]

 - The belkinunv driver now supports Belkin F6C1100-UNV
   [Dave Breiland]

 - The isbmex driver has been upgraded to version 0.05, which fixes
   various errors in formulas, add shutdown capability and revert
   back baudrate to B9600 (instead of B2400), as it broke the
   communication
   [Ricardo Martinezgarza]

 - The support of Sysgration UPGUARDS Pro650 in fentonups has
   been fixed
   [Simon J. Rowe]

 - The packaging files for Red Hat have received various fixes
   [Thomas Jarosch]

 - The solis driver has been fixed to avoid a naming colision and
   compile on Solaris
   [Paweł Kierdelewicz]

 - The snmp-ups driver has corrected the problem when exposing
   certain time data.

---------------------------------------------------------------------------
Release notes for NUT 2.0.2 - what's new since 2.0.1:

 - the newhidups USB driver has been improved a lot and is no more
   experimental. It also now has a basic APC support, which will
   soon replace the legacy hidups driver.

 - The mge-utalk driver has improved its support for old units.

 - The mge-shut driver has been improved for restart/shutdown
   sequences which was previously blocking the serial port.

 - The general MGE support has been added Pulsar EXtreme C / EX RT,

drivers/sms.c
/* sms.c - driver for SMS UPS hardware

   Copyright (C) 2001  Marcio Gomes  <tecnica@microlink.com.br>

   based on fentonups.c:

   Copyright (C) 1999  Russell Kroll <rkroll@exploits.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

   2001/05/17 - Version 0.10 - Initial release
   2001/06/01 - Version 0.20 - Add Battery Informations in driver
   2001/06/04 - Version 0.30 - Updated Battery Volts range, to reflect a correct
                               percent ( % )
   2002/12/02 - Version 0.40 - Update driver to new-model, based on Fentonups 
                               driver version 0.90  
   2002/12/18 - Version 0.50 - Add Sinus Single 2 KVA in database, test with
                               new Manager III sinusoidal versions.
                               Change Detect Name, SMS do not pass real Models
                               in Megatec Info command, only the VA/KVA version
   2002/12/18 - Version 0.51 - Updated Battery Volts range, to reflect a correct
                               percent ( % )
   2002/12/27 - Version 0.60 - Add new UPS Commands SDRET, SIMPWF change BTEST1
   2002/12/28 - Version 0.70 - Add new UPS Commands SHUTDOWN,STOPSHUTD,WATCHDOG
   2003/06/11 - Version 0.71 - Converted to dstate calls and names (rkroll)


   Microlink ISP/Pop-Rio contributed with MANAGER III 1300, MANAGER III 650 UPS
   and Sinus Single 2 KVA for my tests.

   http://www.microlink.com.br and http://www.pop-rio.com.br



*/

#define DRV_VERSION "0.72"

#include "main.h"
#include "serial.h"
#include "sms.h"

#define ENDCHAR 13	/* replies end with CR */

static	int	cap_upstemp = 0;
static	float	lowvolt = 0, voltrange;
static	int	lownorm, highnorm, poll_failures = 0;

static void guessmodel(const char *raw)
{
	char	mch, *mstr;

	mch = raw[17];

        printf ("0         1         2        3         \n");
        printf ("012345678901234567890123567890123456789\n");
        printf ("%s\n", raw);
	mstr = xstrdup(&raw[18]);
	mstr[10] = '\0';	/* 10 chars max, per the protocol */
       
	/* trim whitespace */
	rtrim(mstr, ' ');
     
	dstate_setinfo("ups.model", "SMS %s", mstr);   
	cap_upstemp = 1;

	free(mstr);
}

static void getbaseinfo(void)
{
	char	temp[256], model[32], *raw;
	int	modelnum, i;

	/* dummy read attempt to sync - throw it out */
	ser_send(upsfd, "I\r");
	ser_get_line(upsfd, temp, sizeof(temp), ENDCHAR, "", 3, 0);

	/* now retrieve information and parse */
	ser_send(upsfd, "I\r");
	ser_get_line(upsfd, temp, sizeof(temp), ENDCHAR, "", 3, 0);
	raw = xstrdup(temp);

	if (temp[0] != '#')
		fatalx("Bad UPS info start character [%s]", temp);

	temp[11] = 0;
	temp[27] = 0;

	/* manufacturer */
	rtrim(&temp[1], ' ');

	dstate_setinfo("ups.mfr", "%s", &temp[1]);

/*
        0         1         2        3         
        012345678901234567890123567890123456789
        #SMS LTDA         1300 VA   VER 1.0 
        #SMS LTDA        1300VA SEN VER 5.0 
        #SMS LTDA             2 KVA   VER 1.0 
*/
	/* grab full model string */

	rtrim(&temp[17], ' ');

	snprintf(model, sizeof(model), "%s", &temp[17]);
/*        printf("->%s<-\n",model);  */
	modelnum = -1;

	/* figure out official model name and voltage info from table */
	for (i = 0; modeltab[i].mtext != NULL; i++) {
		if (!strcmp(modeltab[i].mtext, model)) {
			modelnum = i;
			lowvolt = modeltab[i].lowvolt;
			voltrange = modeltab[i].voltrange;
			cap_upstemp = modeltab[i].has_temp;
			break;
		}
	}

	/* table lookup fails -> guess */
	if (modelnum == -1)
		guessmodel (raw);
	else {
		dstate_setinfo("ups.model", "%s", modeltab[modelnum].desc);

		dstate_setinfo("input.transfer.low", "%i", 
			modeltab[modelnum].lowxfer);
		dstate_setinfo("input.transfer.high", "%i", 
			modeltab[modelnum].highxfer);

		lownorm = modeltab[modelnum].lownorm;
		highnorm = modeltab[modelnum].highnorm;
	}

	/* now add instant command support info */
	dstate_addcmd("test.battery.start");
	dstate_addcmd("test.battery.stop");
	dstate_addcmd("shutdown.return");	/* was CMD_SDRET */
	dstate_addcmd("test.failure.start");
	dstate_addcmd("shutdown.stayoff");	/* was CMD_SHUTDOWN */
	dstate_addcmd("shutdown.stop");
	dstate_addcmd("reset.watchdog");

	printf("Detected %s on %s\n", dstate_getinfo("ups.model"), device_path);
	free(raw);

	/* paranoia - cancel any shutdown that might already be running */
	ser_send(upsfd, "C\r");
}

static int instcmd(const char *cmdname, const char *extra)
{
	if (!strcasecmp(cmdname, "test.battery.stop")) {
		ser_send(upsfd, "CT\r");
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "test.battery.start")) {
		ser_send(upsfd, "TL\r");/* start battery test until bat low */
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "test.failure.start")) {
		ser_send(upsfd, "T\r");
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "shutdown.return")) {
		/* shutdown and restart */
		ser_send(upsfd, "C\r");
		ser_send(upsfd, "S.3R0003\r");
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "shutdown.stayoff")) {
		/* shutdown now (one way) */
		ser_send(upsfd, "C\r");
		ser_send(upsfd, "S.3\r");
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "shutdown.stop")) {
		/* Cancel Shutdown  */
		ser_send(upsfd, "C\r");
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "reset.watchdog")) {
		/* WATCHDOG Crontab Function */
		ser_send(upsfd, "C\r");
		ser_send(upsfd, "S05R0003\r");
		return STAT_INSTCMD_HANDLED;
	}

	upslogx(LOG_NOTICE, "instcmd: unknown command [%s]", cmdname);
	return STAT_INSTCMD_UNKNOWN;
}

void upsdrv_initinfo(void)
{
	getbaseinfo();

	upsh.instcmd = instcmd;
	dstate_setinfo("driver.version.internal", "%s", DRV_VERSION);
}

static void pollfail(const char *why)
{
	poll_failures++;

	/* don't spew into the syslog forever */
	if (poll_failures < 3)
		upslogx(LOG_ERR, why);

	return;
}

void upsdrv_updateinfo(void)
{
	char	temp[256], utility[16], loadpct[16], acfreq[16], battvolt[16],
		upstemp[16], pstat[16], outvolt[16];
	int	util, ret;
	double	bvoltp;

	ser_send(upsfd, "Q1\r");

	ret = ser_get_line(upsfd, temp, sizeof(temp), ENDCHAR, "", 3, 0);

	/* sanity checks for poll data */
	if (strlen(temp) < 46) {
		pollfail("Poll failed: short read from UPS");
		dstate_datastale();
		return;
	}

	if (strlen(temp) > 46) {
		pollfail("Poll failed: oversized read from UPS");
		dstate_datastale();
		return;
	}

	if (temp[0] != '(') {
		pollfail("Poll failed: invalid start character");
		dstate_datastale();
		return;
	}

	if (poll_failures > 0)
		upslogx(LOG_NOTICE, "UPS poll succeeded");

	poll_failures = 0;		

	/* (MMM.M NNN.N PPP.P QQQ RR.R S.SS TT.T  b7b6b5b4b3b2b1b0<cr>
	 *
	 * MMM.M : input voltage (utility)
	 * NNN.N : fault voltage (ignored)
	 * PPP.P : output voltage
	 */

	sscanf(temp, "%*c%s %*s %s %s %s %s %s %s", utility, outvolt, loadpct,
		acfreq, battvolt, upstemp, pstat);

	dstate_setinfo("output.voltage", "%s", outvolt);
	dstate_setinfo("input.voltage", "%s", utility);
	dstate_setinfo("battery.voltage", "%s", battvolt);

	bvoltp = ((atof(battvolt) - lowvolt) / voltrange) * 100.0;

	if (bvoltp > 100.0)
		bvoltp = 100.0;

	dstate_setinfo("battery.charge", "%02.1f", bvoltp);

	status_init();

	util = atoi(utility);

	if (pstat[0] == '0') {
		status_set("OL");		/* on line */

		/* only allow these when OL since they're bogus when OB */
		if (pstat[2] == '1') {		/* boost or trim in effect */
			if (util < lownorm)
				status_set("BOOST");

			if (util > highnorm)
				status_set("TRIM");
		}

	} else {
		status_set("OB");		/* on battery */
	}

	if (pstat[1] == '1')
		status_set("LB");		/* low battery */

	status_commit();

	if (cap_upstemp == 1)
		dstate_setinfo("ups.temperature", "%s", upstemp);

	dstate_setinfo("input.frequency", "%s", acfreq);
	dstate_setinfo("ups.load", "%s", loadpct);

	dstate_dataok();
}

/* power down the attached load immediately */
void upsdrv_shutdown(void)
{
	char	temp[256], pstat[32];

	/* basic idea: find out line status and send appropriate command */

	ser_send(upsfd, "Q1\r");
	ser_get_line(upsfd, temp, sizeof(temp), ENDCHAR, "", 3, 0);
	sscanf (temp, "%*s %*s %*s %*s %*s %*s %*s %s", pstat);

	/* on battery: send S01<cr>, ups will return by itself on utility */
	/* on line: send S01R0003<cr>, ups will cycle and return soon */

	ser_send(upsfd, "S01");

	if (pstat[0] == '0') {			/* on line */
		printf("On line, sending shutdown+return command...\n");
		ser_send(upsfd, "R0003");
	} else
		printf("On battery, sending normal shutdown command...\n");

	ser_send_char(upsfd, 13);	/* end sequence */
}

void upsdrv_help(void)
{
}

void upsdrv_makevartable(void)
{
}

void upsdrv_banner(void)
{
	printf("Network UPS Tools - SMS UPS driver %s (%s)\n", 
		DRV_VERSION, UPS_VERSION);
	printf("by Marcio Gomes at Microlink - tecnica@microlink.com.br\n\n");

	upslogx(LOG_WARNING, "This driver is obsolete and has been replaced by the \"megatec\""
	                     " driver. It will be removed somewhere in the near future.");
}

void upsdrv_initups(void)
{
	upsfd = ser_open(device_path);
	ser_set_speed(upsfd, device_path, B2400);
}

void upsdrv_cleanup(void)
{
	ser_close(upsfd, device_path);
}
